File: vue3-front-end/src/models/user/product.ts
// export default interface Product{
//     product_id: number;
//     product_name: string;
//     category: string;
//     brand: string;
//     price: number;
//     stock: number;
//     market_id: number;
//     status: boolean;
// }

export default class Product{
    product_name: string;
    category: string;
    brand: string;
    price: number;
    stock: number;
    status: boolean;
    product_id?: number;
    market_id?: number;

    constructor(
        product_name: string,
        category: string,
        brand: string,
        price: number,
        stock: number,
        status: boolean,
        product_id?: number,
        market_id?: number
    ){
        if(product_id) {
            this.product_id = product_id;
        }
        this.product_name = product_name;
        this.category = category;
        this.brand = brand;
        this.price = price;
        this.stock = stock;
        this.status = status;
        this.market_id = market_id;
    }
}

export interface UpdateProduct {
    product_id: number;
    product_name: string;
    category: string;
    brand: string;
    price: number;
    stock: number;
    market_id: number;
    status: boolean;
}

export interface CreateProduct {
    product_name: string;
    category: string;
    brand: string;
    price: number;
    stock: number;
}

export interface ChangeStatus {
    product_id: number;
    product_name: string;
    category: string;
    brand: string;
    price: number;
    stock: number;
    market_id: number;
    status: boolean;
}

export interface AddToCart {
    user_id: number;
    product_id: number;
    market_id: number;
    quantity: number;
}

File: flask-back-end/main/services/product.py
from flask import jsonify

from main.models._db import delete, save
from main.models.product import Product
from main.schemas.product import ProductSchema

# import datetime

class ProductService:
    def __init__(self):
        self.product_schema = ProductSchema()
        self.products_schema = ProductSchema(many=True)

    def get_all(self):
        products = Product.query.all()
        return jsonify(self.products_schema.dump(products))

    def get(self, product_id):
        product = Product.query.get(product_id)
        return self.product_schema.jsonify(product)

    def create(self, data):
        product = Product.query.filter_by(product_name = data['product_name']).first()
        if not product:
            new_product = Product(
                product_name= data['product_name'],
                category= data['category'],
                brand = data['brand'],
                price = data['price'],
                stock = data['stock'],
                market_id =  data['market_id'],
                status = data['status'],
            )

            save(new_product)
            return self.product_schema.jsonify(new_product)
    
    def get_all_by_market(self, market_id):
        products = Product.query.filter_by(market_id=market_id)
        return self.products_schema.jsonify(products)

    def update(self, product_id, data):
        product = Product.query.get(product_id)
        if product:
            product.product_name = data['product_name'],
            product.category = data['category'],
            product.brand = data['brand'],
            product.price = data['price'],
            product.stock = data['stock'],
            product.status = data['status']
            save(product)
            return self.product_schema.jsonify(product)

    def delete(self, product_id):
        product = Product.query.get(product_id)
        if product:
            delete(product)
            return product_id